sw/device/lib/crypto/drivers/hmac.c

```diff
@@ -8,148 +8,420 @@
-  abs_mmio_write32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_KEY_0_REG_OFFSET,
-                   key->key[0]);
-  abs_mmio_write32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_KEY_1_REG_OFFSET,
-                   key->key[1]);
-  abs_mmio_write32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_KEY_2_REG_OFFSET,
-                   key->key[2]);
-  abs_mmio_write32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_KEY_3_REG_OFFSET,
-                   key->key[3]);
-  abs_mmio_write32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_KEY_4_REG_OFFSET,
-                   key->key[4]);
-  abs_mmio_write32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_KEY_5_REG_OFFSET,
-                   key->key[5]);
-  abs_mmio_write32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_KEY_6_REG_OFFSET,
-                   key->key[6]);
-  abs_mmio_write32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_KEY_7_REG_OFFSET,
-                   key->key[7]);
-
-  // Initialize in HMAC mode.
-  hmac_init(kHardenedBoolTrue);
+/**
+ * Write given byte array into the `MSG_FIFO`. This function should only be
+ * called when HWIP is already running and expecting further message bytes.
+ */
+static void write_to_msg_fifo(const uint8_t *msg, size_t msg_len) {
+  // TODO(#23191): Improve message copying by writing them in 32b at a time.
+  for (size_t i = 0; i < msg_len; i++) {
+    abs_mmio_write8(kHmacBaseAddr + HMAC_MSG_FIFO_REG_OFFSET, msg[i]);
+  }
 }
 
-void hmac_update(const uint8_t *data, size_t len) {
-  // Individual byte writes are needed if the buffer isn't word aligned.
-  for (; len != 0 && (uintptr_t)data & 3; --len) {
-    abs_mmio_write8(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_MSG_FIFO_REG_OFFSET,
-                    *data++);
+status_t hmac_init(hmac_ctx_t *ctx, const hmac_mode_t hmac_mode,
+                   const hmac_key_t *key) {
+  if (ctx == NULL) {
+    return OTCRYPTO_BAD_ARGS;
   }
+  // TODO(#23191) Zeroise or randomly populate ctx struct during init.
+
+  // Prepare cfg_reg in context.
+  ctx->cfg_reg = HMAC_CFG_REG_RESVAL;
+  // The endianness is fixed at driver level and not exposed to the caller.
+  // Digest should be big-endian to match the SHA-256 specification.
+  ctx->cfg_reg =
+      bitfield_bit32_write(ctx->cfg_reg, HMAC_CFG_DIGEST_SWAP_BIT, true);
+  // Message should be little-endian to match Ibex's endianness.
+  ctx->cfg_reg =
+      bitfield_bit32_write(ctx->cfg_reg, HMAC_CFG_ENDIAN_SWAP_BIT, false);
 
-  for (; len >= sizeof(uint32_t); len -= sizeof(uint32_t)) {
-    uint32_t data_aligned = read_32(data);
-    abs_mmio_write32(TOP_EARLGREY_HMAC_BASE_ADDR + HMAC_MSG_FIFO_REG_OFFSET,
-                     data_aligned);
-    data += sizeof(uint32_t);
+  // We need to keep `sha_en` low until context is restored, see #23014.
+  ctx->cfg_reg = bitfield_bit32_write(ctx->cfg_reg, HMAC_CFG_SHA_EN_BIT, false);
+
+  // Default value for `hmac_en` is false, HMAC calls set it to true below.
+  ctx->cfg_reg =
+      bitfield_bit32_write(ctx->cfg_reg, HMAC_CFG_HMAC_EN_BIT, false);
```

Comment:
Are you able to run this test (FPGA or silicon)? `//SW/device/tests/crypto/cryptotest:hmac_kat`

Reply:
I do not have FPGA or silicon at the moment, so I am running custom tests on sim_verilator and on sim_dv:
https://github.com/lowRISC/opentitan/commit/29b6a51a02aed17a553d05a0a86f45923bdf2a49

I actually managed to get my hands on CW340, but I didn't have time to setup hyperdebug. I assume this test is not part of CI then?